### PR TITLE
Timing workarounds

### PR DESCRIPTION
**Pull request by litghost**

### CODE REVIEW

**Comment by mithro** [comment]
I think @acomodi is probably a better reviewer then myself...

**Comment by HackerFoo** [comment]
@litghost Can you add a test showing that the workaround works?

**Comment by litghost** [comment]
> @litghost Can you add a test showing that the workaround works?

I could, but I need to stay focused on debugging the xc7 VPR timing behavior.  I've already spent 1+ day making various PR's, so I'll leave this for later.  Without this change, timing based placement on the xc7 grid simply fails.

File: vpr/src/place/timing_place_lookup.cpp [hunk 1]
```diff
@@ -538,65 +538,49 @@ float delay_reduce(std::vector<float>& delays, e_reducer reducer) {
 
 /* Take the average of the valid neighboring values in the matrix.
  * use interpolation to get the right answer for empty blocks*/
-static float find_neightboring_average(vtr::Matrix<float> &matrix, int x, int y) {
+static float find_neightboring_average(vtr::Matrix<float> &matrix, int x, int y, int max_distance) {
```
Comment (acomodi) [inline vpr/src/place/timing_place_lookup.cpp:541]: Typo in the name of the function? (neighboring)

File: vpr/src/route/check_rr_graph.cpp [hunk 2]
```diff
@@ -473,7 +473,7 @@ void check_rr_node(int inode, enum e_route_type route_type, const DeviceContext&
 
             if (check_for_out_edges) {
                 std::string info = describe_rr_node(inode);
-                VTR_LOG_WARN( "in check_rr_node: %s has no out-going edges.\n", info.c_str());
+                //VTR_LOG_WARN( "in check_rr_node: %s has no out-going edges.\n", info.c_str());
```
Comment (acomodi) [inline vpr/src/route/check_rr_graph.cpp:476]: Maybe add a `FIXME` or something to easily spot this commented warning?

File: vpr/src/place/timing_place_lookup.cpp [hunk 3]
```diff
@@ -613,6 +597,17 @@ static void fix_uninitialized_coordinates(vtr::Matrix<float>& delta_delays) {
     }
 }
 
+static void fill_impossible_coordinates(vtr::Matrix<float>& delta_delays) {
+    //Set any empty delta's to the average of it's neighbours
+    for (size_t delta_x = 0; delta_x < delta_delays.dim_size(0); ++delta_x) {
+        for (size_t delta_y = 0; delta_y < delta_delays.dim_size(1); ++delta_y) {
+            if (delta_delays[delta_x][delta_y] == IMPOSSIBLE_DELTA) {
+                delta_delays[delta_x][delta_y] = find_neightboring_average(delta_delays, delta_x, delta_y, 5);
```
Comment (acomodi) [inline vpr/src/place/timing_place_lookup.cpp:605]: This is needed to have no `IMPOSSIBLE_DELTAs` (by using max distance as 5 in this case), right?

**Review by acomodi — COMMENTED** [review]
LGTM. Probably this will not affect SymbiFlow as placement is not currently timing driven. Have you tested it against some more complex designs (with timing driven placement)?

File: vpr/src/place/timing_place_lookup.cpp (hunk 3 above)
Comment (litghost) [reply vpr/src/place/timing_place_lookup.cpp:605]: Correct.  The ideal case is to have no holes in the delta_delays matrix, but it also doesn't make sense to take an average that too wide.

**Comment by litghost** [comment]
> LGTM. Probably this will not affect SymbiFlow as placement is not currently timing driven. 

This PR was created because enabling timing based placement didn't work, and was an outgrowth of trying to enable timing based placement.

> Have you tested it against some more complex designs (with timing driven placement)?

So far I've tested it against the dram tester and counter, which is are not particular complicated designs.  It places successfully, but the overall design does not close timing.  It is not clear if this is a problem with the placer or the packer.  This PR was simply the first attempt to get timing based placement to run without crashing.

**Comment by acomodi** [comment]
> So far I've tested it against the dram tester and counter, which is are not particular complicated designs. It places successfully, but the overall design does not close timing. It is not clear if this is a problem with the placer or the packer. This PR was simply the first attempt to get timing based placement to run without crashing.

Ok, IMO it is good to merge, but before maybe it could be worth to add a comment in the code above `fill_impossible_coordinates`  pointing to the relative issue.

**Comment by HackerFoo** [comment]
> > @litghost Can you add a test showing that the workaround works?
> 
> I could, but I need to stay focused on debugging the xc7 VPR timing behavior. I've already spent 1+ day making various PR's, so I'll leave this for later. Without this change, timing based placement on the xc7 grid simply fails.

In particular, I've found that such changes can often have unintended results. If you've looked at the results and they make sense, I'm fine with deferring this to a later issue.

**Comment by litghost** [comment]
> > > @litghost Can you add a test showing that the workaround works?
> > 
> > 
> > I could, but I need to stay focused on debugging the xc7 VPR timing behavior. I've already spent 1+ day making various PR's, so I'll leave this for later. Without this change, timing based placement on the xc7 grid simply fails.
> 
> In particular, I've found that such changes can often have unintended results. If you've looked at the results and they make sense, I'm fine with deferring this to a later issue.

No doubt.  These changes are simply to get us off of the starting line.  Without some method to fill the delay matrix, we can only use bounding box placement.  It is possible that bounding box placement is superior to this delay matrix approximation, but that can be explored with a command line option.

**Review by acomodi — APPROVED** [review]

**Pull request closed without merging**